packages/@n8n/permissions/src/types.ee.ts

```diff
@@ -1,53 +1,81 @@
+	sharing: Scope[];
+};
 
-export type ScopeLevel = 'global' | 'project' | 'resource';
-export type GetScopeLevel<T extends ScopeLevel> = Record<T, Scope[]>;
-export type GlobalScopes = GetScopeLevel<'global'>;
-export type ProjectScopes = GetScopeLevel<'project'>;
-export type ResourceScopes = GetScopeLevel<'resource'>;
-export type ScopeLevels = GlobalScopes & (ProjectScopes | (ProjectScopes & ResourceScopes));
+export type ScopeOptions = { mode: 'oneOf' | 'allOf' };
 
-export type MaskLevel = 'sharing';
-export type GetMaskLevel<T extends MaskLevel> = Record<T, Scope[]>;
-export type SharingMasks = GetMaskLevel<'sharing'>;
-export type MaskLevels = SharingMasks;
+export type RoleNamespace = z.infer<typeof roleNamespaceSchema>;
+export type GlobalRole = z.infer<typeof globalRoleSchema>;
+export type AssignableGlobalRole = z.infer<typeof assignableGlobalRoleSchema>;
+export type CredentialSharingRole = z.infer<typeof credentialSharingRoleSchema>;
+export type WorkflowSharingRole = z.infer<typeof workflowSharingRoleSchema>;
+export type ProjectRole = z.infer<typeof projectRoleSchema>;
 
-export type ScopeMode = 'oneOf' | 'allOf';
-export type ScopeOptions = { mode: ScopeMode };
+/** Union of all possible role types in the system */
+export type AllRoleTypes = GlobalRole | ProjectRole | WorkflowSharingRole | CredentialSharingRole;
 
-export type PublicApiKeyResources = keyof typeof API_KEY_RESOURCES;
+type RoleObject<T extends AllRoleTypes> = {
+	role: T;
+	name: string;
+	scopes: Scope[];
+	licensed: boolean;
+};
 
-export type ApiKeyResourceScope<
-	R extends PublicApiKeyResources,
-	Operation extends (typeof API_KEY_RESOURCES)[R][number] = (typeof API_KEY_RESOURCES)[R][number],
-> = `${R}:${Operation}`;
+export type AllRolesMap = {
+	global: Array<RoleObject<GlobalRole>>;
+	project: Array<RoleObject<ProjectRole>>;
+	credential: Array<RoleObject<CredentialSharingRole>>;
+	workflow: Array<RoleObject<WorkflowSharingRole>>;
+};
 
-// This is purely an intermediary type.
-// If we tried to do use `ResourceScope<Resource>` directly we'd end
-// up with all resources having all scopes.
-type AllApiKeyScopesObject = {
-	[R in PublicApiKeyResources]: ApiKeyResourceScope<R>;
+/**
+ * Represents an authenticated entity in the system that can have specific permissions via a role.
+ * @property role - The global role this principal has
+ */
+export type AuthPrincipal = {
+	role: GlobalRole;
 };
 
-export type ApiKeyScope = AllApiKeyScopesObject[PublicApiKeyResources];
+// #region Public API
+type PublicApiKeyResources = keyof typeof API_KEY_RESOURCES;
+
+type ApiKeyResourceScope<
+	R extends PublicApiKeyResources,
+	Operation extends (typeof API_KEY_RESOURCES)[R][number] = (typeof API_KEY_RESOURCES)[R][number],
+> = `${R}:${Operation}`;
 
-export type GlobalRole = 'global:owner' | 'global:admin' | 'global:member';
-export type AssignableRole = Exclude<GlobalRole, 'global:owner'>;
+export type ApiKeyScope = ApiKeyResourceScope<PublicApiKeyResources>;
```

Comment:
This widens the scope by combining all resources with all operations, as mentioned in the comment in the removed code. E.g. this allows `workflow:changeRole`. 

![image](https://github.com/user-attachments/assets/afc982e1-4904-4ab4-9286-225ec5b78073)

We should bring back the intermediary type.

Reply:
Maybe we should also add a type test that makes sure this is not introduced in the future.

Reply:
Thanks for catching this. reverted and added types test.